test/gpu/native/studies/sort/radixSortGpu.chpl

```diff
@@ -15,101 +16,198 @@ module RadixSort {
   // How many bits we sort at once based on the bitsAtATime
   const buckets = 1 << bitsAtATime; // 2^bitsAtATime, ex: 2^4 = 16 = 0b10000
   const bitMask = buckets - 1; // 2^bitsAtATime - 1, ex: 2^4 - 1 = 15 = 0b1111
-
-  proc serialScan(ref prefixSum: [] uint, buckets: int){
+  // The size of the chunks the input array is divided into for parallelism
+  // Each chunk has it's own count table (hence it's own prefix sum)
+  // Which we then intelligently combine together to produce a sorted output
+  config const chunkSize = 128;
+  config const scanChunkSize = 1024;
+  // TODO Choose better defaults for the above two things
+  config const parallelScan = false; // Need to improve this more
+
+  proc serialScan(ref arr: [] uint, in startIdx = arr.domain.low, in endIdx = arr.domain.high){
     // Convert this count array into a prefix sum
     // This is the same as the count array, but each element is the sum of all previous elements
     // This is an exclusive scan
     // Serial implementation
-
+    startIdx = max(startIdx, arr.domain.low);
+    endIdx = min(endIdx, arr.domain.high);
     // Calculate the prefix sum
     var sum : uint = 0;
-    for i in 0..<buckets {
-      var t : uint = prefixSum[i];
-      prefixSum[i] = sum;
+    for i in startIdx..endIdx {
+      var t : uint = arr[i];
+      arr[i] = sum;
       sum += t;
     }
   }
 
-  proc hillisSteeleScan(ref hostPrefixSum: [] uint, hostBuckets: int){
+  proc makePowerOf2(const sz){
+    // Powers of two only have the highest bit set.
+    // Power of two minus one will have all bits set except the highest.
+    // & those two together should give us 0;
+    // Ex 1000 & 0111 = 0000
+    if (sz & (sz - 1)) == 0 then
+      return sz;
+    // Not a power of two, so we pad it out
+    // To the next nearest power of two
+    const log_2_size = numBytes(uint)*8 - BitOps.clz(sz); // Calculates the log quickly
+    // Next highest nerest power of two is
+    const newSz = 1 << log_2_size;
+    return newSz;
+  }
+
+  proc hillisSteeleScan(ref hostArr: [] uint){
     // Hillis Steele Scan
     // This is the same as the count array, but each element is the sum of all previous elements
     // Uses a naive algorithm that does O(nlogn) work
     // Hillis and Steele (1986)
+    // if hostArr.size==0 then return;
+
     on here.gpus[useGpuId]{
-      var buckets = hostBuckets;
-      var prefixSum = hostPrefixSum;
+      // The algorithm only workd for arrays that are size of a power of two.
+      // In case it's not a power of two we pad it out with 0s
+      const size = makePowerOf2(hostArr.size);
+      var arr : [0..<size] uint;
+      arr[hostArr.domain] = hostArr;
+
       var offset = 1;
-      while offset <= buckets {
-          var prefixSumBuffer = prefixSum;
+      while offset < arr.size {
+          var arrBuffer = arr;
```

Comment:
Why did we make this copy here?

Reply:
Because the parallel loop below modifies the `arr` array so we needed the buffer to ensure correctness.

Reply:
Obviously..